Mathlib/Data/Matroid/Circuit.lean

```diff
@@ -166,4 +171,135 @@ lemma restrict_circuit_iff (hR : R ⊆ M.E := by aesop_mat) :
   simp_rw [circuit_iff, restrict_dep_iff, and_imp, dep_iff]
   exact fun hC hCR h ↦ ⟨⟨⟨hC,hCR.trans hR⟩,fun I hI hIC ↦ h hI.1 (hIC.trans hCR) hIC⟩,hCR⟩
 
+/-! ### Fundamental Circuits -/
+
+/-- For an independent set `I` that spans some `e ∉ I`, the unique circuit contained in `I ∪ {e}`.
+Has the junk value `{e}` if `e ∈ I` and `insert e I` if `e ∉ M.closure I`. -/
+def fundCircuit (M : Matroid α) (e : α) (I : Set α) :=
+  insert e (I ∩ (⋂₀ {J | J ⊆ I ∧ e ∈ M.closure J}))
+
+lemma fundCircuit_eq_sInter (he : e ∈ M.closure I) :
+    M.fundCircuit e I = insert e (⋂₀ {J | J ⊆ I ∧ e ∈ M.closure J}) := by
+  rw [fundCircuit, inter_eq_self_of_subset_right]
+  exact sInter_subset_of_mem ⟨Subset.rfl, he⟩
+
+lemma fundCircuit_subset_insert (M : Matroid α) (e : α) (I : Set α) :
+    M.fundCircuit e I ⊆ insert e I :=
+  insert_subset_insert inter_subset_left
+
+lemma fundCircuit_subset_ground (he : e ∈ M.E) (hI : I ⊆ M.E := by aesop_mat) :
+    M.fundCircuit e I ⊆ M.E :=
+  (M.fundCircuit_subset_insert e I).trans (insert_subset he hI)
+
+lemma mem_fundCircuit (M : Matroid α) (e : α) (I : Set α) : e ∈ fundCircuit M e I :=
+  mem_insert _ _
+
+/-- The fundamental circuit of `e` and `X` has the junk value `{e}` if `e ∈ X` -/
+lemma fundCircuit_eq_of_mem (heX : e ∈ X) (heE : e ∈ M.E := by aesop_mat) :
+    M.fundCircuit e X = {e} := by
+  suffices h : ∀ a ∈ X, (∀ I ⊆ X, e ∈ M.closure I → a ∈ I) → a = e by
+    simpa [subset_antisymm_iff, fundCircuit]
+  exact fun f hfX h ↦ h {e} (by simpa) (mem_closure_of_mem' _ rfl)
+
+/-- A version of `Matroid.fundCircuit_eq_of_mem` that applies when `X ⊆ M.E` instead of `e ∈ X`.-/
+lemma fundCircuit_eq_of_mem' (heX : e ∈ X) (hX : X ⊆ M.E := by aesop_mat) :
+    M.fundCircuit e X = {e} := by
+  rwa [fundCircuit_eq_of_mem]
+
+lemma Indep.fundCircuit_circuit (hI : M.Indep I) (hecl : e ∈ M.closure I) (heI : e ∉ I) :
+    M.Circuit (M.fundCircuit e I) := by
+  apply (hI.inter_right _).insert_circuit_of_forall (by simp [heI])
+  · rw [(hI.subset _).closure_inter_eq_inter_closure, mem_inter_iff, and_iff_right hecl,
+      hI.closure_sInter_eq_biInter_closure_of_forall_subset _
+      (by simp (config := {contextual := true}))]
```

Comment:
```suggestion
      (by simp +contextual]
```

Reply:
and I hope that with this change, the `rw` fits on two lines rather than three?

Reply:
I didn't know about this new syntax. Thanks!